### PR TITLE
Use recaptcha plugin in custom component.

### PR DESCRIPTION
**Pull request by oleg-kosarev**
Pull Request for Issue #21643  .
### Steps to reproduce the issue
Use recaptcha plugin in custom component. Use multiple instances of it so each instance has it's own input name.

When you call the recaptcha plugin to validate you can pass the $code to validate instead of the "standard" code from input.

see the function declaration: public function onCheckAnswer($code = null)

e.g.

$alldata = JFactory::getApplication()->input;
$cresponse = $alldata->get('grecaptcharesponse','','string'); //here we have the catcha code to validate

//call captcha plugin to validate code
JPluginHelper::importPlugin('captcha');
$dispatcher = JEventDispatcher::getInstance();

$res = $dispatcher->trigger('onCheckAnswer',$cresponse);

### Testing Instructions
Setting | Value
PHP Built On | Windows NT DESKTOP-I8R3T2F 10.0 build 17134 (Windows 10) AMD64
Database Type | mysql
Database Version | 5.5.45
Database Collation | utf8_general_ci
Database Connection Collation | utf8mb4_general_ci
PHP Version | 7.2.7
Web Server | Microsoft-IIS/10.0
WebServer to PHP Interface | cgi-fcgi
Joomla! Version | Joomla! 3.8.11 Stable [ Amani ] 31-July-2018 14:00 GMT
Joomla! Platform Version | Joomla Platform 13.1.0 Stable [ Curiosity ] 24-Apr-2013 00:00 GMT

### Expected result
true if vaild
false if faild

### Actual result
Always False

### Documentation Changes Required
if params `$code` not empty $respone = $code
if send params get `g-recaptcha-response`  $respone = $input->get('g-recaptcha-response', '', 'string');
else $response = '';

### CODE REVIEW

**Comment by oleg-kosarev** [comment]
@franz-wohlkoenig This error will be fixed in joomla 3.8.12?

**Comment by ghost** [comment]
this Pull Request needs 2 successfully Tests (not by Developer of PR). Than decide Maintaners if merged or not. 

Thats the Standard-Procedere for Pull Requests.

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 4273c5520ae4f9da6cf5752b072fb52b7acc0450<br /><br />Successfully tested with reCaptcha and the new implemented reCaptcha invisible plugins, as follows : 
* Install this patch and the another one : PR #22270 
* Test the code posted in jcomments.ajax.php : 
	$jinput = JFactory::getApplication()->input;
	$post = $jinput->getArray($_POST);
	$code = $post["g_recaptcha_response"]; //here we have the captcha code to validate

	JPluginHelper::importPlugin('captcha', "recaptcha_invisible");
	$dispatcher = JEventDispatcher::getInstance();
	try
	{
		$res = $dispatcher->trigger('onCheckAnswer',$code);
	}
	catch (Exception $e)
	{
		self::showErrorMessage($e->getMessage());
		$response->addScript("grecaptcha.reset()");
		return $response;
	}<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21651">issues.joomla.org/tracker/joomla-cms/21651</a>.</sub>

**Comment by twister65** [comment]
@oleg-kosarev , please add changes to :
https://github.com/joomla/joomla-cms/blob/459d74686d2a638ec51149d7c44ddab8075852be/plugins/captcha/recaptcha_invisible/recaptcha_invisible.php#L127
I did it myself for the test.

File: plugins/captcha/recaptcha/recaptcha.php [hunk 1]
```diff
@@ -122,7 +122,21 @@ public function onCheckAnswer($code = null)
...
-				$response  = $input->get('g-recaptcha-response', '', 'string');
+				$g_recaptcha_response = $input->get('g-recaptcha-response', '', 'string');
+
+				if (!empty($code))
+				{
+					$response = $code;
+				}
+				elseif (!empty($g_recaptcha_response))
+				{
+					$response  = $g_recaptcha_response;
```
Comment (Quy) [inline plugins/captcha/recaptcha/recaptcha.php]: Remove space before `=`

Comment (Quy) [inline plugins/captcha/recaptcha/recaptcha.php]:
```suggestion
					$response = $g_recaptcha_response;
```

File: plugins/captcha/recaptcha/recaptcha.php [hunk 2]
```diff
@@ -122,7 +122,21 @@ public function onCheckAnswer($code = null)
...
+				}
+				else
+				{
+					$response = '';
+				}
+
 				$spam      = ($response === '');
```
Comment (Quy) [inline plugins/captcha/recaptcha/recaptcha.php]:
```suggestion
				$spam = ($response === '');
```

**Comment by ghost** [comment]
@oleg-kosarev can you make the changes @twister65 suggested at https://github.com/joomla/joomla-cms/pull/21651#issuecomment-423743965?

**Comment by ghost** [comment]
I suggest to close this PR due the lack of response by @oleg-kosarev or @twister65 can make the changes? In the meantime i add the label "needs new owner".

**Comment by ghost** [comment]
@twister65 can you apply the changes or should this Pull Request be closed?

**Comment by twister65** [comment]
@franz-wohlkoenig , I think you can close this PR because the user code is not necessary for the implementation of recaptcha:
https://github.com/joomla/joomla-cms/blob/45e43040641edb3359111d1154e8623743a88321/plugins/captcha/recaptcha/recaptcha.php#L131

I'm waiting for a new PR that implements ajax support in the recaptcha plugin.

**Comment by ghost** [comment]
Closed as stated above.
